Selection context

RichardMXt opened the pull request:
*basic selection context
*some node loading and other bug fixes

RichardMXt committed:
Basic selection context
*basic selection context
*some node loading and other bug fixes

RichardMXt committed:
Selection context tab widget
*An info tab widget is shown when an object is selected

RichardMXt committed:
Progress bar

RichardMXt committed:
Fail fast fix

MichaelXt commented on Assets/Player/PlayerHUD2.gd:
```diff
@@ -25,3 +25,15 @@ func _input(event: InputEvent) -> void:
+  
+  if event.is_action_pressed("toggle_building_info_menu"):
+    var tab_container: TabContainer = self.get_node("HBoxContainer/VBoxContainer/TabContainer")
+    var needed_tab_widget_name = event.get_meta("tab_widget_name")
```
target_...

MichaelXt commented on Assets/Player/PlayerHUD2.gd:
```diff
@@ -25,3 +25,15 @@ func _input(event: InputEvent) -> void:
+    var selected_objects: Array = event.get_meta("selected_objects")
+    for tab_widget in tab_container.get_children():
+      if tab_widget.name == needed_tab_widget_name:
+        if tab_widget is BuildingMenuTabWidget:
```
cast using as

RichardMXt committed:
Object type check fix

MichaelXt commented on Assets/World/Behavior/Selectable/Selectable.gd:
```diff
@@ -0,0 +1,59 @@
+        sprite.material.set_shader_parameter("width", 0)
+
+func _ready():
+  if not sprite or not (sprite is AnimatedSprite2D or sprite is Sprite2D):
```
fail fast: assert at the moment sprite is set

MichaelXt commented on Assets/World/Buildings/Lumberjack/LumberjackTent2D.gd:
```diff
@@ -25,7 +25,9 @@ func wait_for_wood():
-    await self.get_tree().create_timer(self.building_data.processing_time).timeout
+    self.production_timer = self.get_tree().create_timer(self.building_data.processing_time)
+    await self.production_timer.timeout
+    production_timer = null
```
self. ?

MichaelXt commented on Assets/World/Buildings/ProductionBuilding2D.gd:
```diff
@@ -13,6 +12,20 @@ var closest_warehouse_path: Array = []
+    var last_value = should_produce
+    should_produce = value
+    if last_value != should_produce and should_produce:
+      production_loop()
```
add comment: fire and forget

MichaelXt commented on Assets/World/Buildings/ProductionBuilding2D.gd:
```diff
@@ -36,54 +47,27 @@ func show_tooltip_animation(amount: int):
-
   var closest_warehouse_path_yet = null
   for warehouse_pos in warehouse_poses:
     var warehouse_path = self.get_node("/root/Main/Pathfinding").carrier_pathfinding.get_path_to_dest(self.position, warehouse_pos)
```
fail fast get_node

MichaelXt commented on Assets/World/Context/GameContextManager.gd:
```diff
@@ -2,16 +2,23 @@ extends Node
       value.is_active = true
       value.context_entered()
-    current_context = value
+      current_context = value
```
shouldn't current_context always be set in this method?

MichaelXt commented on Assets/World/Context/ObjectSelectedContext.gd:
```diff
@@ -0,0 +1,71 @@
+    for object: Selectable in last_selected_objects:
+      object.is_selected = false
+    # check if there are some selected objects
+    if not selected_objects:
```
avoid "not something" checks. It is better be explicit, like: `if array.length > 0`. Sometimes even for bools: `if some_var == true: ...`

MichaelXt commented on Assets/World/Context/ObjectSelectedContext.gd:
```diff
@@ -0,0 +1,71 @@
+      for object: Selectable in selected_objects:
+        object.is_selected = false
+      selected_objects = []
+      %GameContextManager.current_context = null
```
fail fast %GameContextManager

MichaelXt commented on Assets/World/Context/SelectionBox.gd:
```diff
@@ -0,0 +1,27 @@
+    var m_pos := get_viewport().get_mouse_position()
+    if sel_pos_start != m_pos:
+      # the wimdow size is (1920, 1080) and the user might have resized the window
+      draw_line(sel_pos_start, Vector2(m_pos.x, sel_pos_start.y), line_color, line_width / get_viewport().size.y * 1080) # m_pos x
```
-1 for line widths?

MichaelXt commented on Assets/World/Units/Carrier/Carrier.gd:
```diff
@@ -16,6 +16,7 @@ class_name Carrier
 
 func _ready():
+  self.set_up_unit()
   movment_loop()
```
nit: typo: movement

MichaelXt commented on Assets/World/Units/Ships/Ship2D/Ship2D.gd:
```diff
@@ -10,9 +10,10 @@ class_name Ship2D
 signal buoy_added
 
 func _ready():
+  self.set_up_unit()
```
nit: typo: setup

MichaelXt commented on Assets/World/World.gd:
delete

RichardMXt committed:
Some bug fixes
*fail fast fixes
*moved hard functionality from setget to func
*simplified box drawing line width setting
*naming fixes
*variable accessing fix "self."
*added comments
*avoid "not somthing" checks

RichardMXt committed:
Function name typo fix
